src/test/java/teammates/test/cases/datatransfer/FeedbackResponseAttributesTest.java

```diff
@@ -44,4 +48,126 @@ public void testDefaultTimestamp() {
+        assertEquals(Const.TIME_REPRESENTS_DEFAULT_TIMESTAMP, observedFeedbackResponseAttributes.getUpdatedAt());
+    }
+
+    @Test
+    public void testBuilderWithPopulatedFieldValues() {
+        String expectedSessionName = "dummyName";
+        String expectedCourseId = "dummyCourseId";
+        String expectedQuestionId = "dummyQuestionId";
+        FeedbackQuestionType expectedQuestionType = FeedbackQuestionType.TEXT;
+        String expectedGiver = "dummyGiver";
+        String expectedRecipient = "dummyRecipient";
+        String expectedGiverSection = Const.DEFAULT_SECTION;
+        String expectedRecipientSection = Const.DEFAULT_SECTION;
+        Text expectedResponseMetaData = new Text("dummy meta data");
+        String expectedId = expectedQuestionId + "%" + expectedGiver + "%" + expectedRecipient;
+        Instant expectedCreatedAt = Instant.now();
+        Instant expectedUpdatedAt = Instant.now();
+
+        FeedbackResponseAttributes observedFeedbackResponseAttributes = FeedbackResponseAttributes.builder()
+                .withFeedbackResponseId(expectedId)
+                .withFeedbackSessionName(expectedSessionName)
+                .withCourseId(expectedCourseId)
+                .withFeedbackQuestionId(expectedQuestionId)
+                .withFeedbackQuestionType(expectedQuestionType)
+                .withGiver(expectedGiver)
+                .withGiverSection(expectedGiverSection)
+                .withRecipient(expectedRecipient)
+                .withRecipientSection(expectedRecipientSection)
+                .withResponseMetaData(expectedResponseMetaData)
+                .withCreatedAt(expectedCreatedAt)
+                .withUpdatedAt(expectedUpdatedAt)
+                .build();
+
+        assertEquals(expectedId, observedFeedbackResponseAttributes.getId());
+        assertEquals(expectedSessionName, observedFeedbackResponseAttributes.feedbackSessionName);
+        assertEquals(expectedCourseId, observedFeedbackResponseAttributes.courseId);
+        assertEquals(expectedQuestionId, observedFeedbackResponseAttributes.feedbackQuestionId);
+        assertEquals(expectedQuestionType, observedFeedbackResponseAttributes.feedbackQuestionType);
+        assertEquals(expectedGiver, observedFeedbackResponseAttributes.giver);
+        assertEquals(expectedRecipient, observedFeedbackResponseAttributes.recipient);
+        assertEquals(expectedGiverSection, observedFeedbackResponseAttributes.giverSection);
+        assertEquals(expectedRecipientSection, observedFeedbackResponseAttributes.recipientSection);
+        assertEquals(expectedResponseMetaData, observedFeedbackResponseAttributes.responseMetaData);
+        assertEquals(expectedCreatedAt, observedFeedbackResponseAttributes.getCreatedAt());
+        assertEquals(expectedUpdatedAt, observedFeedbackResponseAttributes.getUpdatedAt());
+    }
+
+    @Test
+    public void testBuilderCopy() {
+        String originalGiver = "giver";
```

Comment:
Why are some of the fields inlined and some not and also the disparity in naming?

Reply:
The fields that are not inlined are needed to construct the `originalId` so I didn't inline them.

Reply:
Notice how you repeated withOriginalId twice?

```Java
.withFeedbackResponseId(originalId)
.withFeedbackResponseId("originalFeedbackQuestionId%originalGiver%")
```
In any case, the problem is that `originalId` doesn't actually need to be valid here, many of your other fields aren't quite valid in that they are not sticking to their respective conventions.

So there's no reason to make use of the other fields to create a valid ID because it's never used anyway when creating the corresponding `FeedbackResponse`.

And even if you do want to make it create a valid ID this is error-prone, let's get rid of `withFeedbackResponseId(..)` altogether and set `feedbackResponseId` in `build()`.

Then in `FeedbackSubmissionEditSaveAction`, we don't pass in the ID but instead validate if  `feedbackResponseId` is of the correct format.

Reply:
I tried constructing the id from `build()`, however many tests failed:
https://github.com/bqnguyen94/teammates/blob/7f786a2bc031de2eb2cfb157e0e4308c950abc92/src/main/java/teammates/logic/core/FeedbackResponsesLogic.java#L376-L408
In this function, `EntityDoesNotExistException` was thrown even though `newResponse.getId()` and `responseToUpdate.getId()` gave the same value (which was `agR0ZXN0chYLEhBGZWVkYmFja1F1ZXN0aW9uGAMM` for some reason).

Reply:
Don't refactor it like this :P because we don't know what will be done to the builder.

For example, maybe something like this:

```Java
FeedbackResponseAttributes.builder()
        .withCourseId(computeCourseIdFromRequest()) // depending on how it's named FromRequest is not necessary
        .withFeedbackQuestionType(computeFeedbackQuestionTypeFromRequest(questionIndex))
        ..
        .build();
```

Reply:
I think the test is failing at line 394 where the `feedbackResponseId` is actually `null` when I checked. Can you double check if this is the case? Anyway if it is so it does seem like we can't get rid of `withFeedbackResponseId` in this PR so let's do the following:

```Java
// In legacy data this can be null so the entity is retrieved by a combination of fields,
// see FeedbackResponsesDb#getFeedbackResponseEntityWithCheck(String, String, String)
private String feedbackResponseId;
```

```Java
        // Some legacy data do not have an ID
        if (newResponse.getId() == null) {
            oldResponseEntity = frDb.getFeedbackResponseEntityWithCheck(newResponse.feedbackQuestionId,
                    newResponse.giver, newResponse.recipient);
       ..
```

And also add some information to the tests: https://github.com/TEAMMATES/teammates/pull/8741/commits/de75a458752eaf483c81e68a97832711f5d38e13#r181832009

Anyway, it might also be that only test data have the missing field. Let’s follow that up separately.

Reply:
It's not possible for an entity that exists in the datastore to have a null ID.

Reply:
Ah, I was wondering if it's possible. So it's only test data that has this problem?

Reply:
So it's not about "legacy data" or "test data"; it's about what's being passed into this method - an attributes object created from an entity loaded from the datastore, or a manually constructed attributes object that may not have an ID.

Reply:
I was inferring that the JSON theoretically came from existing data (they did not have IDs). What would have made it possible they had a missing ID though?

Reply:
The ID depends on feedback question ID, which is only generated upon persistence to the datastore. Hence it's not possible to include it in the JSON data.

Reply:
Perhaps not all questions were included in the test data? See `BackDoorLogic#persistDataBundle` and `BackDoorLogic#injectRealIdsIntoResponses` for context.

Reply:
@bqnguyen94 Sorry for my mistake! if @whipermr5 didn't explain some of the details I don't know what would have happened if we didn't check properly later (wrong comments would have went into production 😅 ).

My guess (without seeing how you create the builder) is this:

```Java
public FeedbackResponseAttributes build() {
    // If the ID is not provided, assume that it is a new FeedbackResponse and generate a new ID
    if (feedbackResponseAttributes.feedbackResponseId == null) {
        feedbackResponseAttributes.feedbackResponseId =
                feedbackResponseAttributes.feedbackQuestionId + "%"
                + feedbackResponseAttributes.giver + "%"
                + feedbackResponseAttributes.recipient;
    }
    return feedbackResponseAttributes;
}
```

Can you check if this is the issue you faced?

If this is the case, let's follow up in a separate issue to see if we can get rid of the following:
https://github.com/TEAMMATES/teammates/blob/6157ad411338974355998471d034523472c2d191/src/main/java/teammates/logic/core/FeedbackResponsesLogic.java#L391-L393

Then in the tests, let's just go with the following:
```Java
// The ID format is `feedbackQuestionId%giverEmail%receiver`.
// However, the ID does not change even when the other fields change.
// Therefore there is no need to have the dependency on other fields in this test.
.withFeedbackResponseId("originalId")
```

Reply:
Yeah that's what I wrote in `build()`. So we are moving to having `feedbackResponseId` being generated automatically instead of manually, and enforcing that it cannot be null? Then isn't `feedbackResponseId` quite redundant since we can pinpoint a response with its `questionId` and `giver` and `recipient` anyway?